### PR TITLE
Validate inputs and outputs from custom time formats

### PR DESCRIPTION
**Pull request by aarchiba**
Currently implementing custom time formats is difficult because the jd1 and jd2 values they produce are not validated; and the values they return when asked to export jd1 and jd2 are constrained by peculiar restrictions. This code enforces that jd1 and jd2 are floats or arrays of floats with matching shapes. It also permits, and tests, a variety of peculiar return types.

The main headache on the return types is the issue of "boxing" scalars in zero-dimensional arrays and unboxing them.

### CODE REVIEW

**Comment by aarchiba** [comment]
> I'm not sure I like this - it means slicing, converting, and everything else will slow down because of the extra validation - much of the code counts on `from_jd=True` to be fast.
> 
> That said, I think there is room for improvement: the `TimeFormat.set_jds()` method could actually check that `jd1` and `jd2` are plain arrays with `dtype=float64` and then do the setting; that way, `super().set_jds()` becomes useful.
> 
> Similarly, the base `to_value()` could check that whatever is passed through is an actual array (there are no other constraints on it).

If speed is a concern, do you have benchmarks I can test to see if this makes a significant difference?

**Comment by mhvk** [comment]
No, speed is not my only concern; it is also that this becomes too complicated for my taste. E.g., I'm not sure there is a big benefit to allowing python float for `jd1` and `jd2`.

Note that I don't quite see the problem with subclasses overriding `set_jds` - this code is mostly for new subclasses, so if we change how we use it ourselves and document that one should call `super().set_jds(...)`, I think we're OK (though it will be annoying that the super call fails on older astropy... an argument to make sure anything we do is in place for 4.0, which is LTS).

p.s.  I may have to take a bit of a break of reviewing `Time` issues - I really need to focus on `Quantity` issues...

**Comment by aarchiba** [comment]
> No, speed is not my only concern; it is also that this becomes too complicated for my taste. E.g., I'm not sure there is a big benefit to allowing python float for `jd1` and `jd2`.

I don't like it and didn't want to allow it but a number of tests started to fail when I tried making everything a float64, let alone enforcing array-ness. So our existing test suite (who knows about users?) relies on the existing ability for jd1 and jd2 to be python floats. That's not something that I introduced. Indeed I had planned not to.

> Note that I don't quite see the problem with subclasses overriding `set_jds` - this code is mostly for new subclasses, so if we change how we use it ourselves and document that one should call `super().set_jds(...)`, I think we're OK (though it will be annoying that the super call fails on older astropy... an argument to make sure anything we do is in place for 4.0, which is LTS).

There's quite a lot of difference between asking users to make a non-enforced code change (calling super, and presumably at the end?) and adding checks that the data is already in the right format that are transparent to users. In particular, asking users to call super() will not necessarily make writing custom time formats easier; it will provide another way for them to get it silently wrong. If you're asking them to change all their set_jd methods, why not ask them to call a new method that does the validation I currently have in properties? That's an easy change from my end.

Similarly, if you're going to change how they produce values, why not require them to use to_value instead of value, and have value always have the same implementation (call to_value with no arguments)? Well, okay, this will cause existing unchanged code to fail in mysterious ways, but so does adding a new requirement that they go through super().

> p.s. I may have to take a bit of a break of reviewing `Time` issues - I really need to focus on `Quantity` issues...

No worries. I'll keep filing bugs and submitting PRs as long as I feel like I'm accomplishing something. I should probably get back to PINT and try to do some actual science with it anyway.

**Comment by mhvk** [comment]
`t._time.jd1` ever being a python float now is (yet another?) bug...

**Comment by aarchiba** [comment]
> `t._time.jd1` ever being a python float now is (yet another?) bug...

Ah! In that case I can go squash it everywhere, including the tests that require it. That will simplify this code on both ends.

**Comment by aarchiba** [comment]
> `t._time.jd1` ever being a python float now is (yet another?) bug...

There is one test that checks that (without setting the writeable flag to false) Time objects made from arrays are writeable but those made from scalars are not. If I always box scalars into arrays, this property changes. Should I change the test, set the writeable flag automatically for zero-dimensional arrays, or, uh, maybe just for scalars that aren't arrays?

For now I think I'll just change the test. "All Time objects are writeable or not as reflected by their flag" is easy to explain.

**Comment by aarchiba** [comment]
Oh dear. Writability and broadcasting is a challenge. This may affect code without this PR too. I'd better write a test or two.

**Comment by taldcroft** [comment]
I just looked through this.  In general the design philosophy has been that test our own format classes to ensure compliance with internal requirements on jd1 and jd2, and expect developers of custom formats to do the same.  

Having said that I can see an argument for being a little more helpful, but I do share the concerns of @mhvk about performance and complexity.  The new `_shaped_like_input` is noticeably more difficult to parse (for me at least) in terms of functional work because of being intertwined with validation code.

Maybe my main question is whether it is possible to leave `_shaped_like_input` alone (at least for now) and only include the new validation in the jd1/jd2 setters.  It seems that once the TimeFormat jd1, jd2 are validated and created, then most of the potential issues are caught.  A custom format class has to really mess up to make an output value that would get caught by the new tests in `_shape_like_input`.  In particular passing a function by forgetting to make `value` a property seems too far into the weeds of obscure problems, and is immediately obvious in testing.

The boxing seems reasonable to me, and I like that the scalar case behaves consistent with arrays for being writable.  The new input validation in the jd1/2 setters looks good as well.

File: astropy/time/tests/test_custom_formats.py [hunk 1]
```diff
@@ -89,14 +92,14 @@ def value(self):
             return self.jd1, self.jd2
 
     t = Time.now()
-    with pytest.raises(AttributeError):
+    with pytest.raises(TypeError):
```
Comment (taldcroft) [inline astropy/time/tests/test_custom_formats.py]: Why these changes?

File: astropy/utils/data.py [hunk 2]
```diff
@@ -903,18 +903,7 @@ def get_free_space_in_dir(path):
...
-        retval = ctypes.windll.kernel32.GetDiskFreeSpaceExW(
-                ctypes.c_wchar_p(path), None, None, ctypes.pointer(free_bytes))
-        if retval == 0:
-            raise OSError('Checking free space on {!r} failed '
-                          'unexpectedly.'.format(path))
-        return free_bytes.value
-    else:
-        stat = os.statvfs(path)
-        return stat.f_bavail * stat.f_frsize
+    return shutil.disk_usage(path).free
```
Comment (taldcroft) [inline astropy/utils/data.py]: Unrelated.

Comment (aarchiba) [reply astropy/utils/data.py]: I don't know how that got in there - it's from PR #9182 . I'll clear it out.

**Comment by aarchiba** [comment]
> Having said that I can see an argument for being a little more helpful, but I do share the concerns of @mhvk about performance and complexity. The new `_shaped_like_input` is noticeably more difficult to parse (for me at least) in terms of functional work because of being intertwined with validation code.

I don't really see how there is interleaving? The first part confirms that it's an array, and unboxes it if it's safe to do so. (Only safe to unbox if it's zero-dimensional and not masked.)

The second part de-numpifies various types, so that users get string types and floats instead of the numpy equivalents.

All of its functionality is driven by the needs of existing test cases, with the exception of the explicit checks that jd1/jd2 are arrays. 

> Maybe my main question is whether it is possible to leave `_shaped_like_input` alone (at least for now) and only include the new validation in the jd1/jd2 setters. It seems that once the TimeFormat jd1, jd2 are validated and created, then most of the potential issues are caught. A custom format class has to really mess up to make an output value that would get caught by the new tests in `_shape_like_input`. In particular passing a function by forgetting to make `value` a property seems too far into the weeds of obscure problems, and is immediately obvious in testing.

That test is in there because I made precisely that mistake and debugging it was fairly difficult, particularly in light of the paucity of documentation. The other validation is because I made several other errors while trying to implement a custom format that were also very hard to debug, not least because it was hard to tell whether the problem was upon construction and creation of jd1/jd2 or construction of the return values. I was tempted to put in another stage of validation that confirmed that the return value matched the shape of the jd1/jd2 arrays but considering possible return arrays (for example ymdhms) it was pretty awkward.

Of course most of the things I pass as return values from custom formats are preposterous; the point is that we can handle whatever type the custom format wants to use. 

The old version is much less consistent about unboxing - it fails with a mysterious error unless the returned value is an array type, it unboxes datetimes whether they are masked or not, and it returns float64s and strings rather than numpy versions - unless the custom format provides the non-numpy versions. If you tried to document its behaviour in terms of what values it deals with and what types come out, it would be a complicated business. The new version is actually much simpler in the sense that it has many fewer special cases and it is much less dependent on exactly what the custom format returns as a value.

File: astropy/time/tests/test_custom_formats.py (hunk 1 above)
Comment (aarchiba) [reply astropy/time/tests/test_custom_formats.py]: In this case the raising of AttributeError was a result of the existing implementation and came with an impenetrable message and backtrace. The new code is able to detect the problem and raise whichever exception seems suitable; in this case TypeError seemed appropriate. Only mis-implemented non-functional custom formats - like those I worked through trying to get one working - trigger this exception, so I do not expect to break any code by changing the exception; the previous exception was never chosen, just an artifact of the previous implementation.

File: astropy/time/core.py [hunk 3]
```diff
@@ -874,12 +874,43 @@ def shape(self, shape):
...
+            # zero-dimensional array
+            if (isinstance(value, np.ndarray)
+                    and not value.shape
+                    and not np.ma.is_masked(value)):
+                out = value[()]
+                #out = value[()] if value.dtype.fields else value.item()
+            else:
+                out = value
+        # De-numpify types for the convenience of callers
+        if isinstance(out, np.float64):
```
Comment (taldcroft) [inline astropy/time/core.py]: There are generally no restrictions on the type of `value` produced by a time format. The existing Time API is that scalar values are always converted to a Python-native type where possible via `item()` and we should maintain that.  For test cases consider:
```
    @property
    def value(self):
        return self.jd1.astype(int)

    @property
    def value(self):
        out = self.jd1 + 1j * self.jd2
        return out

    @property
    def value(self):
        out = np.empty(shape=self.jd1.shape, dtype='S10')
        out[()] = b'junk'
        return out

    @property
    def value(self):
        out = np.empty(shape=self.jd1.shape, dtype='O')
        out[()] = {'a': 1}
        return out
```

File: astropy/time/core.py [hunk 4]
```diff
@@ -874,12 +874,43 @@ def shape(self, shape):
...
+            return float(out)
+        elif isinstance(out, np.str_):
+            return str(out)
+        elif callable(out) and not hasattr(self._time.__class__.value, "fget"):
```
Comment (taldcroft) [inline astropy/time/core.py]: The documentation is quite clear that `value` needs to be a property, so I think this test is not appropriate here and ask that you remove it.  A better place to do this validation is when a new format is added to the registry.  Maybe you can use your experience to add other tests at that point?

File: CHANGES.rst [hunk 5]
```diff
@@ -369,6 +369,9 @@ astropy.time
 - Time formats that do not use ``val2`` now raise ValueError instead of
   silently ignoring a provided value. [#9373]
 
+- Custom time formats (implemented in subclasses of ``TimeFormat``) now have
```
Comment (taldcroft) [inline CHANGES.rst]: This applies to all Time formats, not only "custom" ones.  I usually think of "custom" being user-defined.  The PR title should be updated as well.

File: astropy/time/core.py (hunk 3 above)
Comment (aarchiba) [reply astropy/time/core.py]: This is covered by `test_custom_format_can_return_any_scalar`.

Should this include datetime64? The pre-existing test `test_datetime64_no_format` demands that datetime64 be able emerge from a Time object, but `.item()` turns that into a python `datetime`. There's also record dtypes, which lost their field names and accessors when `.item()`ized. (Both of these lead to test failures, so I have special-cased them.)

File: astropy/time/core.py (hunk 4 above)
Comment (aarchiba) [reply astropy/time/core.py]: It's a cleaner check there, thank you.

File: astropy/time/core.py [hunk 6]
```diff
@@ -874,12 +874,43 @@ def shape(self, shape):
                     reshaped.append(val)
 
     def _shaped_like_input(self, value):
-        out = value
-        if value.dtype.kind == 'M':
-            return value[()]
-        if not self._time.jd1.shape and not np.ma.is_masked(value):
-            out = value[()] if value.dtype.fields else value.item()
-        return out
+        if not isinstance(self._time.jd1, np.ndarray):
```
[the remark this answers is not in the document]
Comment (aarchiba) [reply astropy/time/core.py]: I think you're right. I have resisted the temptation to leave an assertion.

**Comment by bsipocz** [comment]
@aarchiba - this needs now a rebase.

**Comment by aarchiba** [comment]
gerebaseerd

File: astropy/time/core.py [hunk 7]
```diff
@@ -418,12 +418,12 @@ def __init__(self, val, val2=None, format=None, scale=None,
                 # check the location can be broadcast to self's shape.
                 self.location = np.broadcast_to(self.location, self.shape,
                                                 subok=True)
-            except Exception:
+            except Exception as e:
```
Comment (taldcroft) [inline astropy/time/core.py]: Generally prefer to avoid 1-letter variable names in new code.  The standard idiom in this module is `err`, but I also use `exc`.

File: astropy/time/formats.py [hunk 8]
```diff
@@ -438,7 +462,7 @@ def set_jds(self, val1, val2):
             raise ScaleValueError("Cannot convert from '{}' epoch scale '{}'"
                                   "to specified scale '{}', got error:\n{}"
                                   .format(self.name, self.epoch_scale,
-                                          self.scale, err))
+                                          self.scale, err)) from err
```
Comment (taldcroft) [inline astropy/time/formats.py]: I didn't know about this syntax, but I like it!

File: astropy/time/formats.py [hunk 9]
```diff
@@ -1437,4 +1461,36 @@ def value(self):
...
+
+
+def _broadcast_writeable(jd1, jd2):
+    if jd1.shape == jd2.shape:
+        return jd1, jd2
+    # When using broadcast_arrays, *both* are flagged with
+    # warn-on-write, even the one that wasn't modified, and
+    # require "C" only clears the flag if it actually copied
+    # anything.
+    s = np.broadcast(jd1, jd2).shape
```
Comment (taldcroft) [inline astropy/time/formats.py]: Please call this `shape` not `s`.

File: astropy/time/formats.py (hunk 8 above)
Comment (aarchiba) [reply astropy/time/formats.py]: It's new. The tracebacks can be long but they can be lifesavers especially when the problem shows up in CI. I haven't systematically gone through and applied it everywhere, but I couldn't resist adding it when I saw an opportunity.

Not relevant to astropy but six has a function raise_from to let you do this even on python2.

**Comment by aarchiba** [comment]
I'm really not sure what's going on with the Windows failures. The reason Windows fails in ways Linux doesn't is because it doesn't have a long double type, so np.longdouble==np.float64. I don't know how that translates into the error we're seeing though. I'll keep looking. It may take several cycles through CI, though, as I don't really have a Windows machine I can try this on. I'll post something when it's sorted.

**Comment by mhvk** [comment]
Since `longdouble` will be allowed in #9361, do you want to just remove those tests? Alternatively, I'm playing with
```
    @pytest.mark.skipif(np.finfo(np.longdouble).eps >= np.finfo(float).eps,
                        reason="long double is the same as float")
```
p.s. The matplotlib errors seem unrelated. A rebase might get rid of those...

**Comment by pllim** [comment]
MPL tests are broken right now, just ignore them. A rebase won't fix it (yet).

**Comment by aarchiba** [comment]
Victory! Happy to squash and rebase if desired (sorry about all the microcommits).

**Comment by bsipocz** [comment]
@aarchiba - 6 commits for 200+ lines is great, so no need to squash.

**Comment by aarchiba** [comment]
No problem. Thanks for putting up with my stubbornness!
